LSPS0/common-schemas.md

```diff
@@ -0,0 +1,296 @@
+* For SegWit v1 (Taproot) or later, encoded using [bech32m][].
+
+[bech32]: https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki
+[bech32m]: https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki
+
+Readers MUST support, and writers SHOULD only emit:
+
+* SegWit v0 addresses with 20 byte commitment (P2WPKH) or
+  32 byte commitment (P2WSH).
+* SegWit v1 addresses with 32 byte commitment (Taproot).
+
+Readers MAY support other SegWit versions.
+
+### Lightning Network Node Signatures
+
+Signatures generated by a particular Lightning Network node, with a
+particular known node ID, MUST be generated and represented using
+the [LND `signmessage` #specinatweet][], and encoded as a JSON string
+containing the zbase32 encoding.
+
+[LND `signmessage` #specinatweet]: https://web.archive.org/web/20191010011846/https://twitter.com/rusty_twit/status/1182102005914800128
+
+> **Rationale** This non-BOLT specification is widely immplemented
+> and is commonly used in many bespoke protocols, often used to
+> prove that a particular user owns a particular node, or as an
+> informal way to reduce spam by requiring that a participant
+> prove they have a published Lightning Network node.
+
+CLN, LND, and Eclair all have a `signmessage` command that allows
+generation of this signature, and verification can be implemented
+using (**Non-normative**) e.g. [ln-verifymessagejs][].
+
+[ln-verifymessagejs]: https://github.com/SeverinAlexB/ln-verifymessagejs
+
+Other LSPS specifications MUST specify that messages to be signed
+include the string `LSPS` followed by the LSPS specification
+number, as well as a human-readable ASCII text warning not to sign
+the message manually.
+For example, a hypothetical LSPS999 might specify:
+
+ ```Markdown
+ The signature MUST sign a message with the template:
+
+     "LSPS999: DO NOT SIGN THIS MESSAGE MANUALLY: I will make an `OP_RETURN` output with ${hash}."
+ ```
+
+> **Rationale** `signmessage` is widely used in
+> many bespoke protocols where a verifier will ask a human operator
+> to prove they control a node by signing a specified message
+> from the verifier.
+> The verifier could then provide a template from an LSPS
+> specification, hoping to get a signature that could be used
+> in an LSPS protocol to prove something other than what the
+> human operator expected.
+
+### Datetimes
+
+Particular points of time in the modern era (a "datetime") MUST be
+encoded as a JSON string containing the [ISO 8601][] format
+`"YYYY-MM-DDThh:mm:ss.uuuZ"`.
+These are always in the UTC timezone.
+
+[ISO 8601]: https://www.iso.org/iso-8601-date-and-time-format.html
+
+### Binary Blobs (Raw Transactions, PSBTs, Lightning onions, etc.)
+
+Binary blobs MUST be encoded as a JSON string containing the
+Base 64 encoding of the binary blob, as described in [RFC 4648
+Section 4][].
+
+[RFC 4648 Section 4]: https://datatracker.ietf.org/doc/html/rfc4648#section-4
+
+Padding characters `=` MUST NOT be used.
```

Comment:
Some base64 parsers don't like it if there's no padding.
Source: https://twitter.com/fiatjaf/status/1558525040374718465

Reply:
If so we should *require* padding, not allow it.

Reply:
Done.